Fix some JAWS bugs

alexstine opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #29526

1. JAWS will now have the ability to use Gutenberg navigation mode and here the blocks announced via Tab.
2.Removed broken speak() function call that was only working part of the time.
3. Remove the wordpress/a11y package. No longer needed.
4.Replace `role="group"` with `role="document"` for better block editing experience in earlier versions of JAWS. 
 
## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

I have tested using NVDA and JAWS on Windows 10. I tested using Voiceover on Mac. All seems to be in working order.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Hopefully some much needed Bug fixes.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Add aria-live region directly to button that way JAWS picks up on dynamic content change. Switch role group to role document.

Commit:
Only apply to Windows.

Commit:
Small tweak for Voiceover not working. Remove isWindows not-needed function.

Commit:
Fix tests.

alexstine commented:
Anyone know why these last 2 e2e tests are failing? I thought I had fixed them all running the tests locally, but maybe I missed something?

youknowriad commented on packages/block-editor/src/components/block-list/use-block-props/index.js:
```diff
@@ -151,7 +151,7 @@ export function useBlockProps( props = {}, { __unstableIsHtml } = {} ) {
 		id: `block-${ clientId }${ htmlSuffix }`,
 		tabIndex: 0,
-		role: 'group',
+		role: 'document',
```
Was trying to read a bit about this change here https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Document_Role it does seem to be a bit impactful. The link seem to suggest that the role is more suited for elements that should be read and not interacted with and this is not the case for blocks as we should be able to navigate inside blocks and interact with them in different ways depending on the block.

Would you mind sharing more about this change, Is this just for the navigation mode? I wonder if it's something that could have a bad impact on edit mode.

alexstine replied:
@youknowriad No, this is to make older versions of JAWS work in edit mode. From my testing, I can't see that it has hurt in anyway how screen readers interact with the block. However, I flagged this for the Accessibility Team to test that way we can be confident that it has been tested on different systems.

youknowriad commented on packages/block-editor/src/components/block-tools/block-selection-button.js:
```diff
@@ -90,13 +80,6 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
-		// To solve it we announce the focus change manually.
-		if ( isWindows() ) {
-			speak( label );
-		}
```
I'm really curious why this has been removed and replaced with the `aria-live` attribute, they are supposed to have the same result. I think you mentioned that `speak` was `undefined` for you, is that right? This would mean that there's probably an issue elsewhere that could have larger impacts.

alexstine replied:
@youknowriad Yes, whenever I did a

`console.log( speak( label ) );`

it came back undefined 90% of the time when it was supposed to be called. I have a feeling this is related to how the React hook is used in combination with useEffect(). If I understand it right, useEffect() will run directly after the component is rendered so if that's the case, speak() may not be callable at this point. Instead, I just added the attribute directly to the button because the dynamic change that occurs with Tab key triggers the aria-live region to announce to screen readers. This is an easy fix in comparison to all the debugging that may have to be done to figure out why wordpress/a11y.speak() is having trouble in this context. It's an easier solution so I think it's fine to go with it.


https://www.freecodecamp.org/news/beginners-guide-to-using-react-hooks/

alexstine replied:
I also tried to create a timeout function there to see if waiting some time would allow me access to speak() more reliably, but no dice on that either.

youknowriad replied:
`speak` function doesn't return anything meaning calling `console.log( speak( label ) );` will always print `undefined` in the console, but this doesn't mean the function didn't work. `speak` works by changing the content of a DOM element with the following class `a11y-speak-polite` at the end of the DOM

alexstine replied:
@youknowriad No, I am certain it didn't work. The only time the speak() function seems to be fired is win tabbing from the title field to the first block or from the post options sidebar after the block list, Shift+Tab. Other than that, just tabbing through the list does not trigger the region. It does not announce or change, you can view this with the web inspector.

Also, I tried changing the 2nd arg to "assertive" and still no luck.

`speak( label, 'assertive' );`

I was just trying with console.log() to see if I could find the issue but it lead to nothing. I am still convinced it may have something to do with useEffect().

All I can say for certain is the speak() function only fires sometimes, it doesn't work all the time. E.g. if you have a list of 10 blocks, it will fire on block 1 and block 10 (Shift+Tab) but nowhere else during the list.

tellthemachines replied:
I played around with this a bit and you're right, @alexstine , the problem is with useEffect. It's only running once, when the component first renders, but it should run every time the label changes. We can fix this by adding the label variable inside the square brackets passed as the second parameter of useEffect. For this to work, we'll also have to move the declaration of the label variable to just before useEffect is called, and we should probably move the declaration of blockType too as that variable is only used in computing the value of label. 

We should still remove the `isWindows()` condition, as this issue is also reproducible on Mac with VoiceOver.

The default "polite" value for `speak()` worked well for me on VoiceOver after these changes.

alexstine replied:
Fixed.

joedolson commented:
Tested this with NVDA/Chrome and NVDA/Firefox, and it worked OK. On Chrome, it triple read every block. That's potentially annoying, but it's not a problem at the scale of what this patch fixes.

tellthemachines reviewed (commented):
Tested with NVDA on Windows 10 and VoiceOver on mac and this is working well!  I left a comment with suggested changes so we can continue using the `speak` function instead of using `aria-live` directly. 

The remaining e2e failures are in places where we're searching for elements by aria role and expecting a block to have a group role, so searching in the whole project for any remaining `role: 'group'` should give you all the instances that still need changing. If you'd like any help with this, let me know!

Commit:
Try using speak() function again with modified code for useEffect().

Commit:
Merge branch 'trunk' of https://github.com/wordpress/gutenberg into fix/jaws-navigation-mode

Commit:
Try to fix up tests.

Commit:
Cleanup code.

alexstine commented:
Can I get a final review on this? I believe I've addressed all concerns and thank you to @tellthemachines for showing me the way with this weird useEffect() bug. Now that label is used as an arg, it only fires speak() when label changes.

Addressing @joedolson feedback:

The only other way to fix the repeating NVDA announcement is to completely restructure how Gutenberg does this. Instead of a dynamically updating button, it should be a list of regular buttons. Sadly, this was a bit of a poor implementation and maybe in the future it can be re-worked. That is a discussion for a bigger issue and different PR. Navigation mode should completely transform the screen and present a list of blocks similar to how the List View works.

Thanks all. 🥇

youknowriad reviewed (approved):
Assuming we're ok with the role change for a11y team. The changes here look good to me.

alexstine commented:
I personally think it works fine with all screen readers I've tested with. It is certainly a big improvement from where it was. 👍

joedolson commented:
Based on my testing, it at worst makes one AT combination a little more annoying, but it fixes an issue in another combination that made the feature unusable. So it's a win.
